README
README
======

Download
----------------------

Das Modul wird Open Source unter der Open Software License (OSL 3.0) angeboten und kann unter folgender URL kostenlos heruntergeladen werden:

http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm

Autor & Copyright
------------------

Copyright by: touchdesign
Chief developer: Christin Gruber

Licenses
--------

Released under Open Software License (OSL 3.0).

INSTALLATION PER MODULINSTALLER
-------------------------------

Wechseln Sie in Ihrem Shopsystem über den "Adminbereich" in den Tab "Module" und klicken Sie auf den
Link "Ein neues Modul hinzufügen". Klicken Sie auf den Button "Durchsuchen" und wählen Sie nun das 
zuvor heruntergeladene Modularchiv "clickandbuy-vX.tar" von Ihrer lokalen Festplatte.
Nach dem Sie das Modularchiv angewählt haben bestätigen Sie bitte die Installation des Modules durch einen 
Klick auf den Button "Dieses Modul hochladen". Nach erfolgreicher Installation des Modules erscheint die 
Meldung "Modul wurde erfolgreich installiert".

AUTOMATISCHE HÄNDLER SCHNELLREGISTRIERUNG (Neukunden)
-----------------------------------------------------

Nach erfolgreicher Installation des ClickandBuy Modules wechseln Sie bitte in die Modul-Konfiguration durch Anwahl 
des soeben installierten Modules "ClickandBuy" in der Modulliste unter der Kategorie "Payment" durch einen Klick 
auf auf den Link "Einstellungen". Folgen Sie nun bitte dem Link "Automatische ClickandBuy Händlerregistrierung" um ein
neues Händlerkonto bei ClickandBuy einzurichten. Nach erfolgreicher Installation werden Sie automatisch in das Shopsystem
zurückgeleitet und das Paymentmodul ClickandBuy wird automatisch mit allen notwendigen Daten vorkonfiguriert.

Ressourcen
----------

[1]: http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm

CHANGELOG
// Version 04

Add optional order confirm page
Update translation
Add secure key

// ...

de.php
<?php

global $_MODULE;
$_MODULE = array();
$_MODULE['<{clickandbuy}prestashop>clickandbuy_70b0a550a62eee947575b0cff63f6928'] = 'ClickandBuy';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_ffecbf0fa2ab2fea921af6b73b2fae47'] = 'Akzeptiere Zahlung mit ClickandBuy';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_69a1a3ad8dd5da6db3c4da838a0cf9c7'] = 'Sind Sie sicher das Sie dieses Modul entfernen möchten?';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_aeb8b782e5b6c7c6921dbbe61045f72a'] = 'ClickandBuy \"transaction link\" wird benötigt.';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_e624f763d9c00705de6944116ee7f7e2'] = 'ClickandBuy \"transaction password\" wird benötigt.';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_8e061e4538bf3121f6fa1cd77c1d13f9'] = 'ClickandBuy \"seller id\" wird benötigt.';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_f4d1ea475eaa85102e2b4e6d95da84bd'] = 'Bestätigen';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_c888438d14855d7d96a2724ee9c306bd'] = 'Einstellungen übernommen';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_bfcf9018affff023282987d6070cd077'] = 'Händlerregistrierung';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_086247a9b57fde6eefee2a0c4752242d'] = 'Um dieses Modul nutzen zu können konfigurieren Sie es bitte über die automatische';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_e721ad19ee4f3c9b232849e9e3b89037'] = 'ClickandBuy Händlerregistrierung';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_f4f70727dc34561dfde1a3c529b6205c'] = 'Einstellungen';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_1baa9a830d4f3ee16765df6cd8b7258a'] = 'ClickandBuy Seller-Id';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_7d3c512924400b581865803b903f8231'] = 'Ihre Seller-Id';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_8e343630cae0b25afe645a1cf0cd8b39'] = 'ClickandBuy Transaktions-Link';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_2c3019b23fa13815a34c611454d6d112'] = 'Ihr Premiumlink';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_c769513c574977559bc23fd075aaf172'] = 'ClickandBuy MD5-Passwort';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_bc224951e472e3fc4ba97383633431b6'] = 'Leer lassen zum deaktivieren';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_031e61f67d50b136d8268ff9a474789a'] = 'ClickandBuy Transaktions-Passwort';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_89cb7a69dd4dc7f09a66e6d67df278e2'] = 'Ihr Transaktions-Passwort';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_d9c1790bf4b830a829bf16f68e8b6146'] = 'ClickandBuy Logo?';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_f983b8bc797e7c35373293580ec761f1'] = 'Ja, zeige das ClickandBuy Logo an (empfohlen)';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_8250275bc9a53f9352037808e28a557e'] = 'Nein, kein Logo anzeigen';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_0a70e894e9d995c4435d6038734883d3'] = 'Zeigt das Logo von ClickandBuy in der linken Spalte des Shops';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_f19e5adfc9b78756fc95e75629d2b3e5'] = 'Bestellbestätigung aktivieren?';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_0b9dcd1b935be31eabfd48f85bffadad'] = 'Ja, der Kunde soll die Bestellung bestätigen (empfohlen)';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_bff8987096b007525cad457fdc371230'] = 'Nein, direkt zur Zahlung weiterleiten';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_21bc9df5acb24864a8a26cd6f58a81c7'] = 'Der Kunde soll die Bestellung vor der Weiterleitung zur Zahlseite besätigen';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_06933067aafd48425d67bcb01bba5cb6'] = 'Update';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_6f0118397671d5cdea6c525a67f4753c'] = 'Automatisch generierte URLs';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_e722d72b214116587e4e6082588df694'] = 'Confirmation-Url (Bestätigungs-Url)';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_d0007a0dcdfd2d49d6eefb52ee85d3e9'] = 'EMS Push Script';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_6a26f548831e6a8c26bfbbd9f6ec61e0'] = 'Hilfe';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_43cb884881c0f06da549972ed413c477'] = '@Link:';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_510db974619d0b8069cf3526f7d09230'] = '@Vendor:';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_fbf3ccc8d02ed2ac6b75b8b47a5dff9b'] = '@Author:';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_170149a92b2369179bf59bbd535d2f03'] = 'Warenkorb ID';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_6e3299a269c0ddd13beacc072d748d6d'] = 'Zahlen mit ClickandBuy';
$_MODULE['<{clickandbuy}prestashop>clickandbuy_aa75009d991d5a1252e2e9977161b608'] = 'Zahlen mit ClickandBuy';
$_MODULE['<{clickandbuy}prestashop>confirmation_2e2117b7c81aa9ea6931641ea2c6499f'] = 'Ihre Bestellung auf';
$_MODULE['<{clickandbuy}prestashop>confirmation_75fbf512d744977d62599cc3f0ae2bb4'] = 'ist vollständig.';
$_MODULE['<{clickandbuy}prestashop>confirmation_cfa0f5d192620d230ccbf3664d9d0610'] = 'Der Gesamtbetrag beträgt';
$_MODULE['<{clickandbuy}prestashop>confirmation_8de637e24570c1edb0357826a2ad5aea'] = 'Wir haben festgestellt das ein Problem bei Ihrer Bestellung aufgetreten ist, bitte wenden Sie sich an unseren';
$_MODULE['<{clickandbuy}prestashop>confirmation_64430ad2835be8ad60c59e7d44e4b0b1'] = 'Kundensupport';
$_MODULE['<{clickandbuy}prestashop>payment_execution_79585e92d2943aba36351667e70ed911'] = 'Zahlung mit ClickandBuy';
$_MODULE['<{clickandbuy}prestashop>payment_execution_f1d3b424cd68795ecaa552883759aceb'] = 'Bestellzusammenfassung';
$_MODULE['<{clickandbuy}prestashop>payment_execution_879f6b8877752685a966564d072f498f'] = 'Ihr Warenkorb ist leer.';
$_MODULE['<{clickandbuy}prestashop>payment_execution_da8cb1215fec7b65e9246f93850d93f4'] = 'Sie haben die Zahlung per ClickandBuy ausgewählt.';
$_MODULE['<{clickandbuy}prestashop>payment_execution_167add1e8662b28f92fd0ab08e3c5ff5'] = 'Der Gesamtbetrag der Bestellung beträgt';
$_MODULE['<{clickandbuy}prestashop>payment_execution_1f87346a16cf80c372065de3c54c86d9'] = '(inkl. MwSt.)';
$_MODULE['<{clickandbuy}prestashop>payment_execution_52f64bc0164b0e79deaeaaaa7e93f98f'] = 'Bitte bestätigen Sie Ihre Bestellung durch klicken \"Bestellung bestätigen\"';
$_MODULE['<{clickandbuy}prestashop>payment_execution_569fd05bdafa1712c4f6be5b153b8418'] = 'Andere Zahlart';
$_MODULE['<{clickandbuy}prestashop>payment_execution_46b9e3665f187c739c55983f757ccda0'] = 'Bestellung bestätigen';


emsPush.php
<?php
/**
 * $Id$
 *
 * clickandbuy Module
 *
 * Copyright (c) 2009 touchdesign
 *
 * @category Payment
 * @version 0.4
 * @copyright 01.12.2009, touchdesign
 * @author Christin Gruber, <www.touchdesign.de>
 * @link http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm
 * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 *
 * Description:
 *
 * Payment module clickandbuy by touchdesign
 *
 * --
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@touchdesign.de so we can send you a copy immediately.
 *
 */

require dirname(__FILE__).'/../../config/config.inc.php';
require dirname(__FILE__).'/clickandbuy.php';

$clickandbuy = new clickandbuy();

if(substr($_SERVER["REMOTE_ADDR"],0,11) != '217.22.128.'){
  die("NOK");
}

if(!isset($_POST['xml']) || empty($_POST['xml'])){
  die("NOK");
}else{
  $xmlData = str_replace("\\" , "" ,$_POST['xml']);
  $transactionDetails.="XMLDATA: $xmlData\n";
  $xml = new SimpleXMLElement($xmlData);  
}

$event['id'] = (string)$xml->GLOBAL->{'event-id'};
$event['crn'] = (string)$xml->GLOBAL->crn;
$event['systemID'] = (int)$xml->GLOBAL->systemID;
$event['cb_datetime'] = (string)$xml->GLOBAL->datetime;
$event['action'] = (string)$xml->BDR->{'bdr-data'}->action;
$event['annotation'] = (string)$xml->BDR->{'bdr-data'}->annotation;
$event['externalBDRID'] = (string)$xml->BDR->{'bdr-data'}->externalBDRID;
$event['bdrID'] = (string)$xml->BDR->{'bdr-data'}->{'bdr-id'};
$event['price'] = (string)$xml->BDR->{'bdr-data'}->price;
$event['currency'] = (string)$xml->BDR->{'bdr-data'}->currency;
$event['linkID'] = (string)$xml->BDR->{'bdr-data'}->{'link-nr'};
 
if(empty($event['id']) || empty($event['action']) 
  || empty($event['externalBDRID'])){
  die("NOK");
}

$cart = new Cart($event['externalBDRID']);
$orderState = $clickandbuy->getEMSOrderState($event['action']);
if($orderState !== false){
  @$clickandbuy->switchOrderState($event['externalBDRID'], $orderState, 
    floatval(number_format($cart->getOrderTotal(true, 3), 2, '.', '')), 
    $clickandbuy->l('TransactionID: '.$event['externalBDRID']).' '.$clickandbuy->l('Action: '.$event['action']));
}

echo "OK";

?>

clickandbuy.php
<?php
/**
 * $Id$
 *
 * clickandbuy Module
 *
 * Copyright (c) 2009 touchdesign
 *
 * @category Payment
 * @version 0.4
 * @copyright 01.12.2009, touchdesign
 * @author Christin Gruber, <www.touchdesign.de>
 * @link http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm
 * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 *
 * Description:
 *
 * Payment module clickandbuy by touchdesign
 *
 * --
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@touchdesign.de so we can send you a copy immediately.
 *
 */

class clickandbuy extends PaymentModule {

  private $_html = '';

  public function __construct() 
  {
    $this->name = 'clickandbuy';
    $this->tab = 'Payment';
    if (version_compare(_PS_VERSION_, '1.4.0', '<')){
      $this->tab = 'Payment';
    }else{
      $this->tab = 'payments_gateways';
    }
    $this->version = '0.4';
    $this->author = 'touchdesign';
    $this->currencies = true;
    $this->currencies_mode = 'radio';
    parent::__construct();
    $this->page = basename(__FILE__, '.php');
    $this->displayName = $this->l('ClickandBuy');
    $this->description = $this->l('Accepts payments by ClickandBuy');
    $this->confirmUninstall = $this->l('Are you sure you want to delete your details?');
  }

  public function install() 
  {
    $sql = "CREATE TABLE "._DB_PREFIX_."clickandbuy(
      id INT UNSIGNED NOT NULL AUTO_INCREMENT,
      status VARCHAR(30) NOT NULL,
      custRef INT NOT NULL,
      BDRID INT NOT NULL,
      externalBDRID VARCHAR(64),
      price INT NOT NULL,
      currency VARCHAR(3) NOT NULL,
      userIP VARCHAR(15) NOT NULL,
      date_submitted TIMESTAMP,
      PRIMARY KEY (id) 
    ) ENGINE=MyISAM default CHARSET=utf8";
    if(!Db::getInstance()->Execute($sql)){
      return false;
    }
    if (!parent::install() || 
      !Configuration::updateValue('CLICKANDBUY_TRANS_LINK', '') || 
      !Configuration::updateValue('CLICKANDBUY_TRANS_PASSWD', '') || 
      !Configuration::updateValue('CLICKANDBUY_MD5_KEY', Tools::passwdGen(20)) ||
      !Configuration::updateValue('CLICKANDBUY_SELLER_ID', '') ||     
      !Configuration::updateValue('CLICKANDBUY_OS_ERROR', 8) ||
      !Configuration::updateValue('CLICKANDBUY_OS_ACCEPTED', 2) ||
      !Configuration::updateValue('CLICKANDBUY_OS_PENDING', 3) ||
      !Configuration::updateValue('CLICKANDBUY_BLOCK_LOGO', 'Y') ||
      !Configuration::updateValue('CLICKANDBUY_CONFIRM_ORDER', 'Y') ||
      !$this->registerHook('payment') ||
      !$this->registerHook('leftColumn') ||
      !$this->registerHook('paymentReturn')){
      return false;
    }

    return true;
  }

  public function uninstall() 
  {
    if (!Configuration::deleteByName('CLICKANDBUY_TRANS_LINK') || 
      !Configuration::deleteByName('CLICKANDBUY_TRANS_PASSWD') ||
      !Configuration::deleteByName('CLICKANDBUY_MD5_KEY') ||      
      !Configuration::deleteByName('CLICKANDBUY_SELLER_ID') ||
      !Configuration::deleteByName('CLICKANDBUY_OS_ERROR') ||
      !Configuration::deleteByName('CLICKANDBUY_OS_ACCEPTED') ||
      !Configuration::deleteByName('CLICKANDBUY_OS_PENDING') ||
      !Configuration::deleteByName('CLICKANDBUY_BLOCK_LOGO') || 
      !Configuration::deleteByName('CLICKANDBUY_CONFIRM_ORDER') ||
      !parent::uninstall()){
      return false;
    }
    $sql = "DROP TABLE "._DB_PREFIX_."clickandbuy";
    if(!Db::getInstance()->Execute($sql)){
      return false;
    }
    return true;
  }

  private function _postValidation() 
  {
    if (Tools::getValue('submitUpdate')){
      if (!Tools::getValue('CLICKANDBUY_TRANS_LINK')){
        $this->_postErrors[] = $this->l('ClickandBuy "transaction link" is required.');
      }
      if (!Tools::getValue('CLICKANDBUY_TRANS_PASSWD')){
        $this->_postErrors[] = $this->l('ClickandBuy "transaction password" is required.');
      }
      if (!Tools::getValue('CLICKANDBUY_SELLER_ID')){
        $this->_postErrors[] = $this->l('ClickandBuy "seller id" is required.');
      }    
    }
  }

  public function getContent() 
  {
    $this->_html .= '<h2>'.$this->displayName.'</h2>';
    if (Tools::isSubmit('submitUpdate')){
      Configuration::updateValue('CLICKANDBUY_TRANS_LINK', Tools::getValue('CLICKANDBUY_TRANS_LINK'));
      Configuration::updateValue('CLICKANDBUY_TRANS_PASSWD', Tools::getValue('CLICKANDBUY_TRANS_PASSWD'));
      Configuration::updateValue('CLICKANDBUY_MD5_KEY', Tools::getValue('CLICKANDBUY_MD5_KEY'));      
      Configuration::updateValue('CLICKANDBUY_SELLER_ID', Tools::getValue('CLICKANDBUY_SELLER_ID'));
      Configuration::updateValue('CLICKANDBUY_BLOCK_LOGO', Tools::getValue('CLICKANDBUY_BLOCK_LOGO'));
      Configuration::updateValue('CLICKANDBUY_CONFIRM_ORDER', Tools::getValue('CLICKANDBUY_CONFIRM_ORDER'));
    } elseif (Tools::getValue('SellerID') && Tools::getValue('LinkURL') 
      && Tools::getValue('TMIPassword')){
      Configuration::updateValue('CLICKANDBUY_TRANS_LINK', Tools::getValue('LinkURL'));
      Configuration::updateValue('CLICKANDBUY_TRANS_PASSWD', Tools::getValue('TMIPassword'));
      Configuration::updateValue('CLICKANDBUY_SELLER_ID', Tools::getValue('SellerID'));
      $this->getSuccessMessage();
    }

    $this->_postValidation();
    if (isset($this->_postErrors) && sizeof($this->_postErrors)){
      foreach ($this->_postErrors AS $err){
        $this->_html .= '<div class="alert error">'. $err .'</div>';
      }
    }elseif(Tools::getValue('submitUpdate') && !isset($this->_postErrors)){
      $this->getSuccessMessage();
    }

    return $this->_displayForm();
  }

  public function getSuccessMessage()
  {
    $this->_html.='
    <div class="conf confirm">
      <img src="../img/admin/ok.gif" alt="'.$this->l('Confirmation').'" />
      '.$this->l('Settings updated').'
    </div>';
  }

  private function getEmsPushScriptUrl() 
  {
    return (Configuration::get('PS_SSL_ENABLED') == 1 ? 'https://' : 'http://') 
      . $_SERVER['HTTP_HOST']._MODULE_DIR_.$this->name.'/emsPush.php';
  }

  private function getConfirmationUrl() 
  {
    return (Configuration::get('PS_SSL_ENABLED') == 1 ? 'https://' : 'http://')
      . $_SERVER['HTTP_HOST']._MODULE_DIR_.$this->name.'/validation.php';
  }

  private function getShopUrl() 
  {
    return (Configuration::get('PS_SSL_ENABLED') == 1 ? 'https://' : 'http://') 
      . $_SERVER['HTTP_HOST'].__PS_BASE_URI__;
  }

  private function getAutoRegisterUrl() 
  {
    $params = array(
      'portalid' => 'touchDesign',
      'shopurl' => $this->getShopUrl(),
      'shopname' => Configuration::get('PS_SHOP_NAME'),
      'configurationurl' => 'http://'
      . $_SERVER['HTTP_HOST'].$_SERVER['REQUEST_URI'],
      'md5key' => Configuration::get('CLICKANDBUY_MD5_KEY'),
    );

    $urlTouchDesign= 'http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm?';
    foreach($params AS $k => $v){
      $urlTouchDesign .= $k.'='.urlencode($v)."&";
    }

    return substr($urlTouchDesign,0,-1);
  }

  private function _displayForm() 
  {
    $this->_html.= '
      <style type="text/css">
      fieldset a {
        color:#0099ff;
        text-decoration:underline;"
      }
      fieldset a:hover {
        color:#000000;
        text-decoration:underline;"
      }
      </style>';

    $this->_html.= '
      <div><img src="'.$this->_path.'logoBig.png" alt="logoBig.png" alt="logoBig.png" title="ClickandBuy" /></div>
      <br /><br />';

    $this->_html.= '
      <fieldset>
        <legend><img src="'.$this->_path.'logo.gif" />'.$this->l('Merchant Registration').'</legend>
        <div>
          '.$this->l('Automatic').' <a href="'.$this->getAutoRegisterUrl().'"><strong>'.$this->l('ClickandBuy Merchant Registration').'</strong></a>.
        </div>
      </fieldset>
      <br /><br />
      <div class="clear"></div>';

    $this->_html.= '
      <form method="post" action="'.$_SERVER['REQUEST_URI'].'">
      <fieldset>
        <legend><img src="'.$this->_path.'logo.gif" />'.$this->l('Settings').'</legend>
        
        <label>'.$this->l('ClickandBuy seller ID?').'</label>
        <div class="margin-form">
          <input type="text" name="CLICKANDBUY_SELLER_ID" value="'.Configuration::get('CLICKANDBUY_SELLER_ID').'" />
          <p>'.$this->l('Your Seller id').'</p>
        </div>
        <div class="clear"></div>

        <label>'.$this->l('ClickandBuy transaction link?').'</label>
        <div class="margin-form">
          <input type="text" name="CLICKANDBUY_TRANS_LINK" value="'.Configuration::get('CLICKANDBUY_TRANS_LINK').'" />
          <p>'.$this->l('Your Premiumlink').'</p>
        </div>
        <div class="clear"></div>       
        
        <label>'.$this->l('ClickandBuy MD5 Password?').'</label>
        <div class="margin-form">
          <input type="text" name="CLICKANDBUY_MD5_KEY" value="'.Configuration::get('CLICKANDBUY_MD5_KEY').'" />
          <p>'.$this->l('Leave it blank for disabling').'</p>
        </div>
        <div class="clear"></div>   
        
        <label>'.$this->l('ClickandBuy transaction password?').'</label>
        <div class="margin-form">
          <input type="text" name="CLICKANDBUY_TRANS_PASSWD" value="'.Configuration::get('CLICKANDBUY_TRANS_PASSWD').'" />
          <p>'.$this->l('Your transaction password').'</p>
        </div>
        <div class="clear"></div>

        <label>'.$this->l('ClickandBuy Logo?').'</label>
        <div class="margin-form">
          <select name="CLICKANDBUY_BLOCK_LOGO">
            <option '.(Configuration::get('CLICKANDBUY_BLOCK_LOGO') == "Y" ? "selected" : "").' value="Y">'.$this->l('Yes, display the logo (recommended)').'</option>
            <option '.(Configuration::get('CLICKANDBUY_BLOCK_LOGO') == "N" ? "selected" : "").' value="N">'.$this->l('No, do not display').'</option>
          </select>
          <p>'.$this->l('Display logo and payment info block in left column').'</p>
        </div>
        <div class="clear"></div>
        
        <label>'.$this->l('Enable order confirmation?').'</label>
        <div class="margin-form">
          <select name="CLICKANDBUY_CONFIRM_ORDER">
            <option '.(Configuration::get('CLICKANDBUY_CONFIRM_ORDER') == "Y" ? "selected" : "").' value="Y">'.$this->l('Yes, let customer confirm orders (recommended)').'</option>
            <option '.(Configuration::get('CLICKANDBUY_CONFIRM_ORDER') == "N" ? "selected" : "").' value="N">'.$this->l('No, redirect direct').'</option>
          </select>
          <p>'.$this->l('Customer have to confirm their order before redirect to clickandbuy').'</p>
        </div>
        <div class="clear"></div>';
        
    $this->_html.= '
        <div class="margin-form clear pspace"><input type="submit" name="submitUpdate" value="'.$this->l('Update').'" class="button" /></div>
      </fieldset>
      </form>
      <div class="clear"></div>';

    $this->_html.= '
      <br /><br />
      <fieldset>
        <legend><img src="'.$this->_path.'logo.gif" />'.$this->l('URLs').'</legend>
        <b>'.$this->l('Confirmation-Url:').'</b><br /><textarea rows=1 style="width:98%;">'.$this->getConfirmationUrl().'</textarea>
        <b>'.$this->l('emsPush-Script:').'</b><br /><textarea rows=1 style="width:98%;">'.$this->getEmsPushScriptUrl().'</textarea>
      </fieldset>';

    $this->_html.= '
      <fieldset class="space">
        <legend><img src="../img/admin/unknown.gif" alt="" class="middle" />'.$this->l('Help').'</legend>   
        '.$this->l('@Link:').' <a target="_blank" href="http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm">www.touchdesign.de</a><br />
        '.$this->l('@Vendor:').' ClickandBuy International Limited<br />
        '.$this->l('@Author:').' <a target="_blank" href="http://www.touchdesign.de/loesungen/prestashop/clickandbuy.htm">touchdesign</a><br />
      </fieldset><br />';

    return $this->_html;
  }

  public function hookPayment($params) 
  {
    global $smarty;
    
    if(Configuration::get('CLICKANDBUY_CONFIRM_ORDER') == "Y"){
      $smarty->assign(array('gateway' => Tools::getHttpHost(true, true).__PS_BASE_URI__.'modules/'.$this->name.'/payment.php'));
    }else{
      $smarty->assign(array('gateway' => $this->getPremiumLink($params['cart'])));
    }
    
    return $this->display(__FILE__, 'clickandbuy.tpl');
  }

  function hookLeftColumn($params) 
  {
    if(Configuration::get('CLICKANDBUY_BLOCK_LOGO') == "N")
      return false;

    return $this->display(__FILE__, 'blockclickandbuylogo.tpl');
  }

  public function hookPaymentReturn($params) 
  {
    global $smarty;

    $orderObj = $params['objOrder'];
    $orderTotal = $params['total_to_pay'];
    $orderCurrency = $params['currency'];
    $currencyObj = $params['currencyObj'];

    if ($orderObj->module != $this->name || !$this->active){
      return false;
    }

    $state = $params['objOrder']->getCurrentState();
    if ($state == Configuration::get('CLICKANDBUY_OS_PENDING') || 
      $state == Configuration::get('CLICKANDBUY_OS_ACCEPTED')){
      $smarty->assign(array(
        'status' => 'accepted',
        'total_to_pay' => Tools::displayPrice($orderTotal, $currencyObj, false, false),
        'id_order' => $params['objOrder']->id
      ));
    }else{
      $smarty->assign('status', 'failed');
    }

    return $this->display(__FILE__, 'confirmation.tpl');
  } 

  public function isPayment() 
  {
    if (!$this->active)
      return false;
    if (!Configuration::get('CLICKANDBUY_TRANS_LINK'))
      return $this->l($this->displayName.' Error: (invalid or undefined transaction link)');
    if (!Configuration::get('CLICKANDBUY_SELLER_ID'))
      return $this->l($this->displayName.' Error: (invalid or undefined transaction link)');

    return true;
  }

  public function switchOrderState($cartId, $orderState, $amount=NULL, $message=NULL) 
  {
    $orderId = Order::getOrderByCartId($cartId);
    if ($orderId){
      $order = new Order($orderId);

      if($order->getCurrentState() != $orderState){
        $history = new OrderHistory();
        $history->id_order = $orderId;
        $history->changeIdOrderState($orderState, $orderId);
        $history->addWithemail(true);    

        if($message !== NULL){
          $orderMessage = new Message();
          $orderMessage->message = $message;
          $orderMessage->private = 1;
          $orderMessage->id_order = $orderId;
          $orderMessage->add();
        }
        
        return 1;
      }
    } else {
      if (version_compare(_PS_VERSION_, '1.4.0', '<')){
        $this->validateOrder($cartId, $orderState, $amount, 
          $this->displayName, $message, null, null, false);
      }else{
        $cart = new Cart($cartId);
        $customer = new Customer($cart->id_customer);
        $this->validateOrder($cartId, $orderState, $amount, 
          $this->displayName, $message, null, null, false, $customer->secure_key);
      }
      return 0;
    }

    return false;
  }

  function getEMSOrderState($action) 
  {
    switch ($action) {
      case 'payment_successful':
      case 'charge back lifted':
      case 'BDR successfully collected from collection agency':
      case 'booked-in':
        return Configuration::get('CLICKANDBUY_OS_ACCEPTED');
      case 'BDR not collected from collection agency':
      case 'booked-out':
      case 'cancelled':
        return Configuration::get('CLICKANDBUY_OS_ERROR');
      case 'BDR to collection agency':
      case 'charge back':
        return Configuration::get('CLICKANDBUY_OS_PENDING');
      default:
        return false;
    }
  }  

  public function execPayment($cart)
  {
    global $smarty;
    
    $smarty->assign(array('gateway' => $this->getPremiumLink($cart), 
      'nbProducts' => $cart->nbProducts(),
      'total' => number_format(Tools::convertPrice($cart->getOrderTotal(), $currency), 2, '.', '')));

    return $this->display(__FILE__, 'payment_execution.tpl');
  }
  
  private function getPremiumLink($cart)
  {
    $isPayment = $this->isPayment();
    if($isPayment !== true){
      return $this->l($isPayment);
    }

    $addressInvoice = new Address(intval($cart->id_address_invoice));
    $addressDelivery = new Address(intval($cart->id_address_delivery));
    $customer = new Customer(intval($cart->id_customer));
    $currency = new Currency(intval($cart->id_currency));
    $countryInvoice = new Country(intval($addressInvoice->id_country));
    $countryDelivery = new Country(intval($addressDelivery->id_country));
    $lang = Language::getIsoById(intval($cart->id_lang));

    if (!Validate::isLoadedObject($addressInvoice) || !Validate::isLoadedObject($addressDelivery) 
      || !Validate::isLoadedObject($customer) || !Validate::isLoadedObject($currency)){
      return $this->l($this->displayName.' Error: (invalid address or customer)');
    }

    $timestamp = time();
    $premiumLink = Configuration::get('CLICKANDBUY_TRANS_LINK');

    $params = array(
      'price' => number_format(Tools::convertPrice($cart->getOrderTotal(), $currency), 2, '.', ''), 
      'cb_currency' => $currency->iso_code, 
      'externalBDRID' => $cart->id,
      'cb_content_name_utf' => $this->l('CartId:').' '.$timestamp.intval($cart->id),
      'cb_content_info_utf' => $customer->firstname.' '.ucfirst(strtolower($customer->lastname)),
      'lang' => $lang,
      'cb_billing_FirstName' => $addressInvoice->firstname,
      'cb_billing_LastName' => $addressInvoice->lastname,
      'cb_billing_Street' => $addressInvoice->address1,
      'cb_billing_City' => $addressInvoice->city,
      'cb_billing_ZIP' => $addressInvoice->postcode,
      'cb_billing_Nation' => $countryInvoice->iso_code,
      'cb_shipping_FirstName' => $addressDelivery->firstname,
      'cb_shipping_LastName' => $addressDelivery->lastname,
      'cb_shipping_Street' => $addressDelivery->address1,
      'cb_shipping_City' => $addressDelivery->city,
      'cb_shipping_ZIP' => $addressDelivery->postcode,
      'cb_shipping_Nation' => $countryDelivery->iso_code,
    );

    $query="";
    foreach($params AS $k => $v){
      $query .= $k."=".urlencode($v)."&";
    }

    $url = $premiumLink.'?'.substr($query,0,-1);
    if(Configuration::get('CLICKANDBUY_MD5_KEY')){
      $url.='&fgkey='.md5(Configuration::get('CLICKANDBUY_MD5_KEY') . "/" . basename($url));
    }
    
    return $url;
  }
}

?>